File: js/leetCodeEasy/pascalsTriangle.js
// Pascals Triangle
// https://leetcode.com/explore/interview/card/top-interview-questions-easy/99/others/601/

/*
Given a non-negative integer numRows,
generate the first numRows of Pascal's triangle.

Example:
Input: 5
Output:
[
     [1],
    [1,1],
   [1,2,1],
  [1,3,3,1],
 [1,4,6,4,1]
]
*/

export const generate = function(numRows) {
    if(numRows === 0){
        return [];
    }
    if(numRows === 1){
        return [[1]];
    }

    const result = [[1]];
    for(let i=1; i<numRows; i++){
        const last = result[i-1]
        const next = [1];
        for(let j=1; j<last.length; j++){
            next.push(last[j-1] + last[j])
        }
        next.push(1)
        result.push(next)
    }

    return result;
};

File: go/leetcode_easy/pascals_triangle/pascals_triangle.go
package pascals_triangle

func Generate(numRows int) [][]int {
	if numRows == 0 {
		return [][]int{{}}
	}

	if numRows == 1 {
		return [][]int{{1}}
	}

	result := [][]int{{1}}

	for i := 1; i < numRows; i++ {
		last := result[i-1]
		next := []int{1}

		for j := 1; j < len(last); j++ {
			next = append(next, last[j-1]+last[j])
		}

		next = append(next, 1)
		result = append(result, next)
	}

	return result
}
